File: src/shared/api/favorites.js
import { invoke } from '@tauri-apps/api/core'
import { restoreLastFocus } from '@shared/api/window'

// 分页查询收藏列表
export async function getFavoritesHistory(params = {}) {
  const { offset = 0, limit = 50, groupName, search, contentType } = params

  const invokeParams = { offset, limit }
  if (groupName && groupName !== '全部') invokeParams.groupName = groupName
  if (search) invokeParams.search = search
  if (contentType) invokeParams.contentType = contentType

  return await invoke('get_favorites_history', invokeParams)
}

// 获取收藏总数
export async function getFavoritesTotalCount(groupName = null) {
  return await invoke('get_favorites_total_count', {
    groupName: groupName === '全部' ? null : groupName,
  })
}

// 添加收藏
export async function addFavorite(title, content, groupName = '全部') {
  const result = await invoke('add_quick_text', { title, content, groupName })
  await invoke('emit_quick_texts_updated')
  return result
}

// 更新收藏
export async function updateFavorite(id, title, content, groupName) {
  const result = await invoke('update_quick_text', { id, title, content, groupName })
  await invoke('emit_quick_texts_updated')
  return result
}

// 删除收藏
export async function deleteFavorite(id) {
  return await invoke('delete_quick_text', { id })
}

// 移动收藏项位置（拖拽排序）
export async function moveFavoriteItemById(groupName, fromId, toId) {
  return await invoke('move_favorite_item_by_id', {
    groupName: groupName === '全部' ? null : groupName,
    fromId,
    toId,
  })
}

// 粘贴收藏内容
export async function pasteFavorite(id, format = null) {
  try {
    await restoreLastFocus()
    const params = { favorite_id: id }
    if (format) {
      params.format = format
    }

    await invoke('paste_content', { params })

    // 检查是否启用一次性粘贴
    const { getToolState } = await import('@shared/services/toolActions')
    const isOneTimePasteEnabled = getToolState('one-time-paste-button')

    if (isOneTimePasteEnabled) {
      try {
        await deleteFavorite(id)
        const { refreshFavorites } = await import('@shared/store/favoritesStore')
        setTimeout(async () => {
          await refreshFavorites()
        }, 200)
      } catch (deleteError) {
        console.error('一次性粘贴：删除收藏项失败', deleteError)
      }
    }

    return true
  } catch (error) {
    console.error('粘贴收藏内容失败:', error)
    throw error
  }
}

File: src-tauri/src/commands/favorites.rs
use crate::services::database::{
    query_favorites, get_favorites_count,move_favorite_by_id,
    add_clipboard_to_favorites as db_add_clipboard_to_favorites,
    move_favorite_to_group as db_move_favorite_to_group,
    delete_favorite as db_delete_favorite,
    get_favorite_by_id,
    add_favorite as db_add_favorite,
    update_favorite as db_update_favorite,
    FavoritesQueryParams, PaginatedResult, FavoriteItem
};
use crate::services::paste::FilesData;
use std::path::Path;

fn fill_file_exists_for_favorites(items: &mut [FavoriteItem]) {
    for item in items.iter_mut() {
        if item.content_type == "file" || item.content_type == "image" {
            check_and_fill_file_exists(item);
        }
    }
}

fn check_and_fill_file_exists(item: &mut FavoriteItem) {
    if !item.content.starts_with("files:") { return; }
    
    if let Ok(mut data) = serde_json::from_str::<FilesData>(&item.content[6..]) {
        for file in &mut data.files {
            let actual_path = resolve_stored_path(&file.path);
            file.exists = Path::new(&actual_path).exists();
            file.actual_path = Some(actual_path);
        }
        if let Ok(json) = serde_json::to_string(&data) {
            item.content = format!("files:{}", json);
        }
    }
}

fn resolve_stored_path(stored_path: &str) -> String {
    crate::services::resolve_stored_path(stored_path)
}

// 分页查询收藏列表
#[tauri::command]
pub fn get_favorites_history(
    offset: Option<i64>,
    limit: Option<i64>,
    group_name: Option<String>,
    search: Option<String>,
    content_type: Option<String>,
) -> Result<PaginatedResult<FavoriteItem>, String> {
    let params = FavoritesQueryParams {
        offset: offset.unwrap_or(0),
        limit: limit.unwrap_or(50),
        group_name,
        search,
        content_type,
    };
    
    let mut result = query_favorites(params)?;
    fill_file_exists_for_favorites(&mut result.items);
    Ok(result)
}

// 获取收藏总数
#[tauri::command]
pub fn get_favorites_total_count(group_name: Option<String>) -> Result<i64, String> {
    get_favorites_count(group_name)
}

// 移动收藏项（拖拽排序）
#[tauri::command]
pub fn move_favorite_item_by_id(
    group_name: Option<String>,
    from_id: String,
    to_id: String,
) -> Result<(), String> {
    move_favorite_by_id(group_name, from_id, to_id)
}

// 从剪贴板历史添加到收藏
#[tauri::command]
pub fn add_clipboard_to_favorites(id: i64, group_name: Option<String>) -> Result<FavoriteItem, String> {
    db_add_clipboard_to_favorites(id, group_name)
}

// 移动收藏项到分组
#[tauri::command]
pub fn move_quick_text_to_group(id: String, group_name: String) -> Result<(), String> {
    db_move_favorite_to_group(id, group_name)
}

// 删除收藏项
#[tauri::command]
pub fn delete_quick_text(id: String) -> Result<(), String> {
    db_delete_favorite(id)
}

// 根据 ID 获取单个收藏项
#[tauri::command]
pub fn get_favorite_item_by_id_cmd(id: String) -> Result<FavoriteItem, String> {
    get_favorite_by_id(&id)?
        .ok_or_else(|| format!("收藏项不存在: {}", id))
}

// 添加收藏项
#[tauri::command]
pub fn add_quick_text(title: String, content: String, group_name: Option<String>) -> Result<FavoriteItem, String> {
    db_add_favorite(title, content, group_name)
}

// 更新收藏项
#[tauri::command]
pub fn update_quick_text(id: String, title: String, content: String, group_name: Option<String>) -> Result<FavoriteItem, String> {
    db_update_favorite(id, title, content, group_name)
}

// 复制收藏项内容（不记录到历史）
#[tauri::command]
pub fn copy_favorite_item(id: String) -> Result<(), String> {
    use crate::services::paste::set_clipboard_from_item;
    
    let item = get_favorite_by_id(&id)?
        .ok_or_else(|| format!("收藏项不存在: {}", id))?;
    
    set_clipboard_from_item(&item.content_type, &item.content, &item.html_content, true)
}

